src/function/scalar/list/list_lambdas.cpp

```diff
@@ -0,0 +1,417 @@
+			sel.set_index(elem_cnt, source_idx);
+
+			// for each column, set the index of the selection vector to slice properly
+			for (idx_t col_idx = 0; col_idx < args.ColumnCount() - 1; col_idx++) {
+				sel_vectors[col_idx].set_index(elem_cnt, indexes[col_idx]);
+			}
+			elem_cnt++;
+		}
+	}
+
+	DataChunk input_chunk;
+	DataChunk lambda_chunk;
+	input_chunk.InitializeEmpty(types);
+	lambda_chunk.Initialize(result_types);
+
+	ExecuteExpression(types, result_types, elem_cnt, sel, sel_vectors, input_chunk, lambda_chunk, child_vector, args,
+	                  expr_executor);
+	auto &lambda_vector = lambda_chunk.data[0];
+
+	if (IS_TRANSFORM) {
+		AppendTransformedToResult(lambda_vector, elem_cnt, result);
+	} else {
+		AppendFilteredToResult(lambda_vector, result_entries, elem_cnt, result, curr_list_len, curr_list_offset,
+		                       appended_lists_cnt, lists_len, curr_original_list_len, input_chunk);
+	}
+}
+
+static void ListTransformFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	ListLambdaFunction<>(args, state, result);
+}
+
+static void ListFilterFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	ListLambdaFunction<false>(args, state, result);
+}
+
+static void TransformExpression(unique_ptr<Expression> &original, unique_ptr<Expression> &replacement,
+                                ScalarFunction &bound_function, vector<unique_ptr<Expression>> &arguments,
+                                LogicalType &list_child_type) {
+
+	// check if the original expression is a lambda parameter
+	bool is_lambda_parameter = false;
+	if (original->expression_class == ExpressionClass::BOUND_COLUMN_REF) {
+
+		// determine if this is the lambda parameter
+		auto &bound_col_ref = (BoundColumnRefExpression &)*original;
+		if (bound_col_ref.binding.table_index == DConstants::INVALID_INDEX) {
+			is_lambda_parameter = true;
+		}
+	}
+
+	if (is_lambda_parameter) {
+		// this is a lambda parameter, so the replacement refers to the first argument, which is the list
+		replacement = make_unique<BoundReferenceExpression>(arguments[0]->alias, list_child_type, 0);
+
+	} else {
+		// this is not a lambda parameter, so we need to create a new argument for the arguments vector
+		replacement = make_unique<BoundReferenceExpression>(original->alias, original->return_type, arguments.size());
+		bound_function.arguments.push_back(original->return_type);
+		arguments.push_back(move(original));
+	}
+}
+
+static void IterateChildren(ScalarFunction &bound_function, vector<unique_ptr<Expression>> &arguments,
+                            LogicalType &list_child_type, unique_ptr<Expression> &expr) {
+
+	if (expr->expression_class == ExpressionClass::BOUND_SUBQUERY) {
+		throw InvalidInputException("Subqueries are not supported in lambda expressions!");
+	}
+
+	// these expression classes do not have children, transform them
+	if (expr->expression_class == ExpressionClass::BOUND_CONSTANT ||
```

Comment:
`ExpressionIterator::EnumerateChildren` will just not do anything for these expressions - is it necessary to have this big `if` here?

Reply:
I believe so, yes. Because for all of these cases we need to push back/adjust the arguments vector to contain the correct columns for execution.